Adds a new concept page for autoscaling

mocdaniel opened the pull request:
This PR introduces a new Concept page Autoscaling, providing an overview of the different options available. It fixes #44084.

I plan to add to the skeleton currently pushed to the PR over the next days.

Some questions for now:

- Is the overall style acceptable? I read the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/), but finding a balance of formatting too much/too little is hard
- Is a single file `concepts/workloads/autoscaling.md` fine or should it rather go into a new directory, e.g. `concepts/workloads/autoscaling/_index.md`?
- How detailed should the overview be? For example, the [Horizontal Pod Autoscaling Task](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) is incredibly detailed - should I link there **OR** copy bits and pieces **OR** will content be moved over in the future? The HPA overview is one paragraph + a link to the corresponding task for now

---
[Preview](https://deploy-preview-44959--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/autoscaling/)

dipesh-rawat commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+and periodically adjusts the number of {{< glossary_tooltip text="replicas" term_id="replica" >}}
+in a workload to match observed resource utilization such as CPU or memory usage.
+
+There is a [walkthrough example](../../../tasks/run-application/horizontal-pod-autoscale-walkthrough.md) of configuring a HorizontalPodAutoscaler for a Deployment.
```
Some early review of the draft content, the current way of specifying the link is incorrect and leads to a 404 error.  Consider using suggested way below.

```suggestion
There is a [walkthrough example](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough) of configuring a HorizontalPodAutoscaler for a Deployment.
```




Additionally, you can also verify how this content renders by checking the build preview link: [deploy-preview/docs/concepts/workloads/autoscaling/](https://deploy-preview-44959--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/autoscaling/)

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+  With Autoscaling, you can automatically update your workloads or infrastructure in one way or another. This allows your cluster to react to changes in resource demand more elastically and efficiently.
+content_type: concept
+weight: 40
+hide_summary: true # Listed separately in section index
```
```suggestion
```
unless you edit the section index to make that true

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+<!-- overview -->
+
+The concept of _Autoscaling_ in Kubernetes refers to the ability to automatically update the available
+resources of your cluster. This can be either a replication controller (for example a
+{{< glossary_tooltip text="Deployment" term_id="deployment" >}} or
+{{< glossary_tooltip text="ReplicaSet" term_id="replica-set" >}}), or the cluster infrastructure
+itself (for example the number of {{< glossary_tooltip text="Nodes" term_id="node" >}}).
```
Avoid using the term “replication controller”; it's likely to mislead. We'd now say something like “an object that lets you manage a set of Pods”.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+
+_tba_ about VerticalPodAutoscaler
+
+## Scaling the Cluster
+
+_tba_ about Cluster Autoscaler and Karpenter
```
Given the title of “autoscaling workloads”, a discussion about node autoscaling belongs in a different page, I think. It's definitely OK to include a _summary_ of the topic in this page.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+reviewers: []
+title: Autoscaling Workloads
+description: >-
+  With Autoscaling, you can automatically update your workloads or infrastructure in one way or another. This allows your cluster to react to changes in resource demand more elastically and efficiently.
```
“autoscaling” or “automatic scaling” please.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+---
+reviewers: []
```
```suggestion
```

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+
+<!-- body -->
+
+## Scaling Workloads Horizontally
```
We write headings (not page titles, but headings) in [sentence case](https://kubernetes.io/docs/contribute/style/style-guide/#headings)

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+The first option is referred to as _horizontal scaling_, while the second is referred to as _vertical scaling_.
+
+<!-- body -->
+
```
You might like to discuss manual scaling first: you can either change the scale of an object, or use [in-place resizing](https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/).

Choose whether you want to explain both of those before you talk about any automatic scaling, or if you'd like to split _Scaling workloads horizontally_ into manual and then automatic (and the same for vertical scaling).

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+
+_tba_ about Cluster Autoscaler and Karpenter
+
+## Advanced Scenarios
```
I'd make a heading for each kind of scaling (and avoid the term “advanced”):

- Event driven automatic scaling
- Sizing a workload to match your cluster
- Scheduled scaling
  (this is for things like turning off a workload at weekends)

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,55 @@
+
+## Scaling Workloads Vertically
+
+_tba_ about VerticalPodAutoscaler
```
Also mention in-place manual autoscaling.

sftim reviewed (commented):
Thanks. Here's an early review.

sftim commented:
What this page will really benefit from: diagrams. I can help out with some tips.

sftim commented:
> How detailed should the overview be?

Let's start with “not very detailed”.

mocdaniel commented:
Thanks for the incredible feedback, I'll incorporate the quick wins tonight and will continue structuring the docs page according to your suggestions afterwards.

mocdaniel commented:
> What this page will really benefit from: diagrams. I can help out with some tips.

I'm always in favor of visuals. Anything on your mind in particular? Which tool is preferred for diagrams? DrawIO? Excalidraw? Mermaid?



> > How detailed should the overview be?
> 
> Let's start with “not very detailed”.

Gotcha. I resolved all remarks, added a short overview of manual scaling, and rewrote the outline of the remainder of the page. Will continue with vertical scaling tomorrow.

sftim commented:
We have a [diagram guide](https://kubernetes.io/docs/contribute/style/diagram-guide/), and an associated blog [article](https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/).

You can use either Mermaid, or anything else rendered to SVG.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+## Scaling workloads manually
+
+Kubernetes supports _manual scaling_ of workloads, either by changing the number of
+{{< glossary_tooltip text="Replicas" term_id="replica">}} of an object that manages a set of
```
(nit)
```suggestion
{{< glossary_tooltip text="replicas" term_id="replica">}} define for an object that manages a set of
```

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+{{< glossary_tooltip text="Pods" term_id="pod" >}} (for example a {{< glossary_tooltip text="Deployment" term_id="deployment" >}}),
+or by adjusting the provided resources of each Replica (for example CPU or memory).
+
+### Scaling the number of Replicas of a workload
```
```suggestion
### Manual changes to replica count
```

As this is a a concept pages, you're defining concepts (like the idea that you can manually change replica count).
Separately, we could make a task page about this - perhaps under https://kubernetes.io/docs/tasks/run-application/ - that reiterates how to run `kubectl scale`.

mocdaniel replied:
I need some clarification on this - do you want me to not go into detail regarding how to (manually) scale workloads at all, or do you want me to adjust the section headers to make it sound less of a tutorial, but keep the short paragraphs below?

For these paragraphs, I took a look at the [Deployment concept page](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) which also gives short outlines of e.g. how to create deployments using a manifest.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
````diff
@@ -0,0 +1,104 @@
+You can use the `kubectl scale` command to increase or decrease the number of Replicas of a workload:
+
+```shell
+kubectl scale deployment <deployment-name> --replicas=<desired-replicas>
````
(nit)
```suggestion
kubectl scale deployment <deployment-name> --replicas=<desired-count>
```

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+
+### Scaling the number of Replicas of a workload
+
+You can use the `kubectl scale` command to increase or decrease the number of Replicas of a workload:
```
```suggestion
You can use the `kubectl scale` command to increase or decrease the number of replicas for a workload:
```

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+
+### Resizing workloads in-place
+
+Instead of scaling the number of Replicas of a workload, you can also adjust the provided resources
+of each Replica, in-place. You do this by patching the entries in one or both of the following
+fields of the `Pod` or [PodTemplate](/docs/concepts/workloads/pods/#pod-templates) you want to resize:
```
```suggestion
Instead of scaling the number of replicas of a workload, you can also adjust the provided resources
for a particular pod or set of pods, in-place. You do this by patching the entries in one or both of the following
fields of the `Pod` or [pod template](/docs/concepts/workloads/pods/#pod-templates) you want to resize:
```

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+- `spec.containers[*].resources.requests`
+- `spec.containers[*].resources.limits`
+
+{{< caution >}}
+Resizing a workload in-place **without** restarting the Pods or its Containers requires Kubernetes version 1.27 or later.
+{{< /caution >}}
```
Use a `note`. `caution` is for advice you should follow or you put your cluster at risk.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+Kubernetes also supports _automatic scaling_ of workloads, which is the focus of this page.
+
+The concept of _Autoscaling_ in Kubernetes refers to the ability to automatically update the workloads of your cluster. This can be either an object that manages a set of Pods (for example a
+{{< glossary_tooltip text="Deployment" term_id="deployment" >}} or Pods or PodTemplates themselves.
```
AFAIK, you can't autoscale a bare Pod, nor a PodTemplate.

mocdaniel replied:
Ah, misleading wording on my side. I meant 'Pods or PodTemplates' as in 'even when scaling deployments vertically, its the PodTemplate's specs within the deployment that are being updated'.
Will improve this part.

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+The concept of _Autoscaling_ in Kubernetes refers to the ability to automatically update the workloads of your cluster. This can be either an object that manages a set of Pods (for example a
+{{< glossary_tooltip text="Deployment" term_id="deployment" >}} or Pods or PodTemplates themselves.
+
+Depending on _what_ is being scaled, there are also different options for _how_ to scale:
+
+- scale the number of available instances (such as Replicas in a Deployment)
+- scale the available resources on existing instances themselves (such as CPU or memory of a {{< glossary_tooltip text="Container" term_id="container" >}} in a Pod)
+
+The first option is referred to as _horizontal scaling_, while the second is referred to as _vertical scaling_.
```
Consider introducing this explanation before the automatic scaling section. Manual horizontal scaling is still horizontal scaling.

I'd consider expanding the page intro to cover this detail, or make it the first heading (eg “Vertical and horizontal scaling”).

mocdaniel replied:
Good point, will do. 
I was hesitant to put more stuff in the `overview` part since the page content guidelines tell you to [only use a single paragraph for the overview section](https://kubernetes.io/docs/contribute/style/page-content-types/#concept).

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+In Kubernetes, you can automatically scale a workload horizontally using a _HorizontalPodAutoscaler_ (HPA).
+
+It is implemented as a Kubernetes API resource and a {{< glossary_tooltip text="controller" term_id="controller" >}}
+and periodically adjusts the number of {{< glossary_tooltip text="Replicas" term_id="replica" >}}
```
```suggestion
and periodically adjusts the number of {{< glossary_tooltip text="replicas" term_id="replica" >}}
```

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+and periodically adjusts the number of {{< glossary_tooltip text="Replicas" term_id="replica" >}}
+in a workload to match observed resource utilization such as CPU or memory usage.
+
+There is a [walkthrough example](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough) of configuring a HorizontalPodAutoscaler for a Deployment.
```
```suggestion
There is a [walkthrough tutorial](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) of configuring a HorizontalPodAutoscaler for a Deployment.
```

(a separate PR should move that tutorial into the tutorials section; anyway, it _is_ a tutorial).

mocdaniel replied:
Right. Do you want me to open a separate issue for someone to do this?

sftim commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,104 @@
+- `spec.containers[*].resources.limits`
+
+{{< caution >}}
+Resizing a workload in-place **without** restarting the Pods or its Containers requires Kubernetes version 1.27 or later.
```
Do you, as a cluster administrator, need to ensure that any feature gates are enabled? (I think you do, `InPlacePodVerticalScaling`).

mocdaniel replied:
Yup, it's an alpha staged feature flag. How do I hint at that? By adding text to the `note` shortcode or by putting a `feature-state` shortcode in the top of the concept page?

sftim replied:
I'd add a subheading, and follow the subheading with a feature state shortcode.

sftim reviewed (commented):
More feedback.

If you want to save the custom autoscaling part for a later PR, that's fine by me.

sftim commented:
This work is tangentially relevant to https://github.com/kubernetes/website/issues/45074 (about node lifecycle)

mocdaniel commented:
@sftim I finished the first draft of the whole concept page.

I decided against mentioning the KNative autoscaler as I felt I would have to add too much information about unrelated things on top of it (KNative, FaaS, etc.). Also, I don't really know a lot about KNative to begin with.

Now would be a good time to talk about diagrams again. Did you have anything specific in mind? I thought about adding one showcasing the difference between horizontal and vertical scaling visually, but couldn't come up with any other ideas.

sftim commented:
> diagrams

I'll dig out something - are you on Kubernetes' Slack?

sftim commented:
A subproject we could mention: https://github.com/kubernetes-sigs/cluster-proportional-vertical-autoscaler

mocdaniel commented:
> > diagrams
> 
> I'll dig out something - are you on Kubernetes' Slack?

Yes, I'm `@dbodky`.

> A subproject we could mention: https://github.com/kubernetes-sigs/cluster-proportional-vertical-autoscaler

I'll add  this tonight.

sftim commented:
Great!
/lgtm

reylejano commented on content/en/docs/concepts/workloads/autoscaling.md:
```diff
@@ -0,0 +1,146 @@
+
+Kubernetes supports _manual scaling_ of workloads. Horizontal scaling can be done
+using the `kubectl` CLI.
+For vertical scaling, you need to _patch_ the resource definition of your workload.
```
should there be a note about VPA ? It's an extra install so I would say no but it is part of k/autoscaler
https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler

mocdaniel replied:
If I don't get you wrong, there is a whole paragraph about the VPA:

<img width="706" alt="Screenshot 2024-02-16 at 17 26 51" src="https://github.com/kubernetes/website/assets/28064429/03705bec-340f-4352-9ce1-1aa21785905e">

Or are you referring to something else entirely?

sftim replied:
https://k8s.io/docs/concepts/workloads/autoscaling/#scaling-workloads-vertically

reylejano commented:
Since KEDA is also mentioned, we should also add [VPA](https://kubernetes.slack.com/archives/C0191RDKHU1/p1708099351573539) 
This can be done in a follow-up PR
/approve
